### ISSUE TITLE
Sky flickers in mojo shell on certain devices

### ISSUE THREAD

**Issue by tonygentilcore** [domokit/mojo#298]
I have a private sky demo (can give repro instructions offline) that has severe UI flicker on a 2013 Nexus 7, moderate UI flicker on a OnePlus One, but is rock solid on an Android One (sprout).

Sometimes the flicker just flashes by itself and sometimes various components or pieces of components snap in or out of view upon touch input. I suspect we need to reduce a test case, but I'm not sure the best way to go about that. I also have no idea whether this repros in sky shell.

/cc @abarth

**Comment by abarth** [comment]
The next step here is to build a reduced test case of the flicker.

**Comment by tonygentilcore** [comment]
Repros on @jimbeveridge 's Nexus 5 too. He's getting some GL OOM messages in the log that might be a clue.

**Comment by jimbeveridge** [comment]
Here's a snip from the list of errors. Note the Out of Memory error buried in the list.

```
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
[ERROR:texture_manager.cc(1708)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_OUT_OF_MEMORY : glTexImage2D: 
[ERROR:gles2_cmd_decoder.cc(3432)] [GroupMarkerNotSet(crbug.com/242999)!:54BCCAA1]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete
```

**Comment by tonygentilcore** [comment]
By the way, this does seem like a recent regression (~1 week ago). I never noticed it on the FN based version of that UI (although so much has been changing that I definitely can't say this was introduced by our conversion to Widgets).

**Comment by tonygentilcore** [comment]
@abarth and I confirmed this is specific to mojo shell (sky shell is fine) and readily reproduces via:
```
./mojo/tools/mojo_shell.py --sky sky/sdk/example/stocks/lib/main.dart --android
```

/cc @viettrungluu-cr @MitchRudominer

**Comment by MitchRudominer** [comment]
Since @abarth is probably the person who best understands the relevant differences between Sky Shell and Mojo Shell, could he hazard a guess as to the layer of technology that might be a likely culprit?

**Comment by abarth** [comment]
The difference is that Sky Shell has Skia talking directly to Android GL driver whereas Mojo Shell has Skia talking to a command buffer and the Surfaces service.  My guess is that there's some bad interaction between Skia and the command buffer, potentially something that causes a GPU resource leak.

**Comment by tonygentilcore** [comment]
Sounds hairy. No one seems to have touched that code in many months. Is there anyone outside of @jamesr and @sky2 that stand a chance at debugging this? If not, should we consider coercing mojo to go down the same path as sky (at least temporarily until someone owns the surfaces service)?

/cc @qsr @chrismckillop

**Comment by chrismckillop** [comment]
@iansf has been digging into the re-architecting of this part of the system.  Perhaps he can debug it a little in the short term to see if can't plaster over the issue until the bigger work is done?

**Comment by iansf** [comment]
I agree this seems like something I might be able to figure out.  Tony, do
you want to get me set up on your app sometime today so that I can dive
into it?

**Comment by tonygentilcore** [comment]
Thanks! No need to set that up. See repro instructions in https://github.com/domokit/mojo/issues/298#issuecomment-119759384 .

**Comment by iansf** [comment]
Cool -- I have the repro on my HTC 1+.  Sadly, the driver doesn't log any opengl errors for me, so I will try to get my hands on a Nexus 5.

**Comment by iansf** [comment]
Blocked by #318

**Comment by iansf** [comment]
- I am able to reproduce the graphical errors on a Nexus 5, but I do not get any OpenGL errors being logged on that device.
 - I figured out how to get the Android Monitor GLTracer to work with mojo apps approximately, but for some reason it only records the egl commands, and not the opengl commands (two eglMakeCurrent calls and an eglSwapBuffers call), even though it successfully records the framebuffer contents when they get into a broken state.
 - I turned on SK_DEBUG, GR_GL_CHECK_ERROR, and GR_GL_LOG_CALLS in Skia to see what would happen (one at a time and together).  In all three cases, there are no errors reported, and the graphical corruption stops happening.  This makes me think there is a timing problem with talking to the graphics driver, since the app runs so slowly in those modes that inkwells usually don't even show up.
 - I have not found a working way to attach gdb to a sky app launched with mojo_shell -- even manually calling the adb shell commands gives me problematic errors that I haven't yet found a way around.  At this point, though, I'm not sure that gdb would help me diagnose this problem.

**TL;DR**: I am blocked on this bug.  We need better tools for dealing with graphical corruption issues on Android (or if we have them already, I would love to know about them).

If someone knows how to get a proper GLTrace out of Monitor for mojo apps, that would be a huge help.  @cdotstout -- any experience with this?

**Comment by abarth** [comment]
@iansf There's a way to make the command buffer log all the GL commands it touches.  I'd like around in //gpu/command_buffer for something that looks like logging.

**Comment by cdotstout** [comment]
I don't have experience getting gl traces from mojo apps.  

Have you tried the qualcomm gpu profiler?

https://developer.qualcomm.com/software/adreno-gpu-profiler

**Comment by Hixie** [comment]
This issue was moved to domokit/sky_engine#36
